# SVM Model

In [2]:
import pandas as pd
import numpy as np


from sklearn import svm
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report


from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import LabelEncoder

In [3]:
# Loading given training dataset

In [6]:
train_data=pd.read_csv('/content/SalaryData_Train(1).csv')
train_data

age          workclass  ...          native  Salary
0       39          State-gov  ...   United-States   <=50K
1       50   Self-emp-not-inc  ...   United-States   <=50K
2       38            Private  ...   United-States   <=50K
3       53            Private  ...   United-States   <=50K
4       28            Private  ...            Cuba   <=50K
...    ...                ...  ...             ...     ...
30156   27            Private  ...   United-States   <=50K
30157   40            Private  ...   United-States    >50K
30158   58            Private  ...   United-States   <=50K
30159   22            Private  ...   United-States   <=50K
30160   52       Self-emp-inc  ...   United-States    >50K

[30161 rows x 14 columns]

<b> DATA PREPROCESSING </b>

In [7]:
train_data.columns

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [8]:
labelencoder = LabelEncoder()
train_data.iloc[:, -1] = labelencoder.fit_transform(train_data.iloc[:,-1])
train_data

age          workclass  ...          native  Salary
0       39          State-gov  ...   United-States       0
1       50   Self-emp-not-inc  ...   United-States       0
2       38            Private  ...   United-States       0
3       53            Private  ...   United-States       0
4       28            Private  ...            Cuba       0
...    ...                ...  ...             ...     ...
30156   27            Private  ...   United-States       0
30157   40            Private  ...   United-States       1
30158   58            Private  ...   United-States       0
30159   22            Private  ...   United-States       0
30160   52       Self-emp-inc  ...   United-States       1

[30161 rows x 14 columns]

In [9]:
train_data_x=pd.get_dummies(train_data.iloc[:,:13])
train_data_x

age  educationno  ...  native_ Vietnam  native_ Yugoslavia
0       39           13  ...                0                   0
1       50           13  ...                0                   0
2       38            9  ...                0                   0
3       53            7  ...                0                   0
4       28           13  ...                0                   0
...    ...          ...  ...              ...                 ...
30156   27           12  ...                0                   0
30157   40            9  ...                0                   0
30158   58            9  ...                0                   0
30159   22            9  ...                0                   0
30160   52            9  ...                0                   0

[30161 rows x 102 columns]

In [10]:
x=train_data_x
y=train_data['Salary']


In [11]:
# Splitting data into training and testing data set

In [11]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.3)

In [12]:
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((21112, 102), (21112,), (9049, 102), (9049,))

## Tuning svm model with different values of parameters 'gmma' , 'c' and kernal='rbf' for best accuracy

In [15]:
?GridSearchCV

In [13]:
model = SVC()
param_grid = [{'kernel':['rbf'],'gamma':[50,5,10,0.5],'C':[0.1] }]
gsv = GridSearchCV(model,param_grid)
gsv.fit(x_train,y_train)

GridSearchCV(cv=None, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=None,
             param_grid=[{'C': [0.1], 'gamma': [50, 5, 10, 0.5],
                          'kernel': ['rbf']}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [ ]:
# Accoriding to above gridsearch method these are the best parameters

In [14]:
gsv.best_params_ , gsv.best_score_ 

({'C': 0.1, 'gamma': 0.5, 'kernel': 'rbf'}, 0.7494789592039173)

In [ ]:
# Building model with above parameteres

In [15]:
model_2=SVC(C=0.01,gamma=0.5,kernel='rbf')
model_2.fit(x_train , y_train)
y_pred = model_2.predict(x_test)
acc = accuracy_score(y_test, y_pred) * 100
print("Accuracy =", acc)

Accuracy = 75.52215714443585


We got accurcy of 75.5% for the model with C=0.01,gamma=0.5 and kernal as rbf


## Building model with c=0.01 and gamma as default






In [16]:
model_final = SVC(C=0.01,kernel='rbf')
model_final.fit(x_train , y_train)

SVC(C=0.01, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [17]:
y_pred = model_final.predict(x_test)

In [18]:
confusion_matrix(y_test, y_pred)

array([[6813,   21],
       [1817,  398]])

In [19]:
acc = accuracy_score(y_test, y_pred) * 100
print("Accuracy =", acc)

Accuracy = 79.68836335506685


With above parameters we got better accuracy of 79.6% than compared with  previous model. 

## Applying Prediction using above model for given test data

In [21]:
data_test=pd.read_csv('/content/SalaryData_Test(1).csv')
data_test

age      workclass      education  ...  hoursperweek          native  Salary
0       25        Private           11th  ...            40   United-States   <=50K
1       38        Private        HS-grad  ...            50   United-States   <=50K
2       28      Local-gov     Assoc-acdm  ...            40   United-States    >50K
3       44        Private   Some-college  ...            40   United-States    >50K
4       34        Private           10th  ...            30   United-States   <=50K
...    ...            ...            ...  ...           ...             ...     ...
15055   33        Private      Bachelors  ...            40   United-States   <=50K
15056   39        Private      Bachelors  ...            36   United-States   <=50K
15057   38        Private      Bachelors  ...            50   United-States   <=50K
15058   44        Private      Bachelors  ...            40   United-States   <=50K
15059   35   Self-emp-inc      Bachelors  ...            60   United-States    >50K

[15060 rows x 14 columns]

In [23]:
labelencoder = LabelEncoder()
data_test.iloc[:, -1] = labelencoder.fit_transform(data_test.iloc[:,-1])
data_test

age      workclass      education  ...  hoursperweek          native Salary
0       25        Private           11th  ...            40   United-States      0
1       38        Private        HS-grad  ...            50   United-States      0
2       28      Local-gov     Assoc-acdm  ...            40   United-States      1
3       44        Private   Some-college  ...            40   United-States      1
4       34        Private           10th  ...            30   United-States      0
...    ...            ...            ...  ...           ...             ...    ...
15055   33        Private      Bachelors  ...            40   United-States      0
15056   39        Private      Bachelors  ...            36   United-States      0
15057   38        Private      Bachelors  ...            50   United-States      0
15058   44        Private      Bachelors  ...            40   United-States      0
15059   35   Self-emp-inc      Bachelors  ...            60   United-States      1

[15060 rows x 14 columns]

In [24]:
data_test_x=pd.get_dummies(data_test.iloc[:,:13])
data_test_x

age  educationno  ...  native_ Vietnam  native_ Yugoslavia
0       25            7  ...                0                   0
1       38            9  ...                0                   0
2       28           12  ...                0                   0
3       44           10  ...                0                   0
4       34            6  ...                0                   0
...    ...          ...  ...              ...                 ...
15055   33           13  ...                0                   0
15056   39           13  ...                0                   0
15057   38           13  ...                0                   0
15058   44           13  ...                0                   0
15059   35           13  ...                0                   0

[15060 rows x 102 columns]

In [26]:
X_test=data_test_x
Y_test=data_test['Salary']

In [27]:
Y_test_pred=model_final.predict(X_test)

In [31]:
confusion_matrix(Y_test, Y_test_pred)

array([[11319,    41],
       [ 3017,   683]])

In [32]:
acc = accuracy_score(Y_test, Y_test_pred) * 100
print("Accuracy =", acc)

Accuracy = 79.6945551128818


<b> For the given test dataset we got accuracy of 79.7% using SVM model we created </b>